prusti-encoder/src/encoders/type/predicate.rs

```diff
@@ -61,22 +61,37 @@ pub enum PredicateEncData<'vir> {
 #[derive(Clone, Debug)]
 pub struct PredicateEncOutputRef<'vir> {
     /// Constructs the Viper predicate application.
-    pub ref_to_pred: PredicateIdent<'vir, UnaryArity<'vir>>,
-    /// Construct snapshot from Viper predicate.
-    pub ref_to_snap: FunctionIdent<'vir, UnaryArity<'vir>>,
+    pub ref_to_pred: PredicateIdent<'vir, UnknownArity<'vir>>,
+    /// Construct snapshot from Viper ref.
+    pub ref_to_snap: FunctionIdent<'vir, UnknownArity<'vir>>,
     /// Construct snapshot from an unreachable.
     pub unreachable_to_snap: FunctionIdent<'vir, NullaryArity<'vir>>,
-    /// Ref as first argument, snapshot as second. Ensures predicate
-    /// access to ref with snapshot value.
-    pub method_assign: MethodIdent<'vir, BinaryArity<'vir>>,
+    /// Ref as first argument, followed by type parameters, followed by
+    /// snapshot. Ensures predicate access to ref with snapshot value. This
+    /// probably shouldn't be accessed directly, instead see
+    /// `RustTyPredicatesEncOutputRef::apply_method_assign`.
+    pub (super) method_assign: MethodIdent<'vir, UnknownArity<'vir>>,
     /// Always `TypeData::Domain`.
     pub snapshot: vir::Type<'vir>,
     //pub method_refold: &'vir str,
     pub specifics: PredicateEncData<'vir>,
+    pub generics: &'vir [vir::LocalDecl<'vir>],
 }
 impl<'vir> task_encoder::OutputRefAny for PredicateEncOutputRef<'vir> {}
 
 impl<'vir> PredicateEncOutputRef<'vir> {
+    pub fn ref_to_args<'tcx>(
+        &self,
+        vcx: &'vir vir::VirCtxt<'tcx>,
```

Comment:
Generally, I think we should avoid passing `vcx`: it can be acquired inside the function with `with_vcx`.

Reply:
This function is called within a call to `with_vcx` sometimes, or where there's already a vcx in context, e.g

```rust
let inner_ref_to_args = inner.ref_to_args(
    self.vcx,
    lifted_ty,
    self_ref
);
```

Is there not much overhead / problems with nesting `with_vcx` calls? If so, I can remove this (and some other places that take an explicit vcx parameter).